lib/containers/runtime.pm

```diff
@@ -131,18 +222,178 @@ sub remove_image {
     $self->_rt_assert_script_run("rmi -f $image_name");
 }
 
+=head2 remove_container
+
+Remove a container from the pool.
+
+=cut
+sub remove_container {
+    my ($self, $container_name) = @_;
+    $self->_rt_assert_script_run("rm -f $container_name");
+}
+
+=head2 check_image_in_host_registry
+
+Returns true if host contains C<img> or false.
+
+=cut
+sub check_image_in_host_registry {
+    my ($self, $img) = @_;
+    my @lregistry = $self->enum_images();
+    $img =~ @lregistry;
+}
+
+=head2 setup_registry
+
+Updates the registry files for the running container runtime to allow access to
+insecure registries.
+
+Implementation is subject to the the subclass otherwise call to this subroutine dies.
+
+=cut
+sub setup_registry {
+    my ($self) = shift;
+    die "Unsupported runtime - " . $self->runtime;
+}
+
 =head2 cleanup_system_host
 
-Remove containers and then all the images respectively and then make sure that everything was cleaned up.
+Remove containers and then all the images respectively.
+Asserts that everything was cleaned up unless c<assert> is set to 0.
 
 =cut
 sub cleanup_system_host {
+    my ($self, $assert) = @_;
+    $assert // 1;
+    $self->_rt_assert_script_run("ps -q | xargs -r " . $self->runtime . " stop", 180);
+    $self->_rt_assert_script_run("system prune -a -f",                           180);
+
+    if ($assert) {
+        assert_equals(0, scalar @{$self->enum_containers()}, "containers have not been removed");
+        assert_equals(0, scalar @{$self->enum_images()},     "images have not been removed");
+    }
+}
+
+# Container runtime subclasses here
+package containers::runtime::docker;
+use Mojo::Base 'containers::runtime';
+use testapi;
+use containers::utils qw(registry_url);
+use utils qw(systemctl file_content_replace);
+has runtime => 'docker';
+
+sub setup_registry {
     my ($self) = shift;
-    # copy from common > clean_container_host
-    $self->_rt_assert_script_run("stop \$($self->{runtime} ps -q)", 180) if script_output("$self->{runtime} ps -q | wc -l") != '0';
-    $self->_rt_assert_script_run("system prune -a -f",              180);
-    assert_equals(0, scalar @{$self->enum_containers()}, "containers have not been removed");
-    assert_equals(0, scalar @{$self->enum_images()},     "images have not been removed");
+    my $registry = registry_url();
+    # Allow our internal 'insecure' registry
+    assert_script_run(
+        'echo "{ \"debug\": true, \"insecure-registries\" : [\"localhost:5000\", \"registry.suse.de\", \"' . $registry . '\"] }" > /etc/docker/daemon.json');
+    assert_script_run('cat /etc/docker/daemon.json');
+    systemctl('restart docker');
+    record_info "setup $self->runtime", "deamon.json ready";
+}
+
+sub up {
```

Comment:
This function is exactly the same as the super class... can you help me understand why you do redefine it here too?

Reply:
there was some experiments between docker, podman etc. might be leftover. need to check

Reply:
*this change is ready for review*